patroni/ctl.py

```diff
@@ -1372,3 +1373,222 @@ def format_pg_version(version):
         return "{0}.{1}.{2}".format(version // 10000, version // 100 % 100, version % 100)
     else:
         return "{0}.{1}".format(version // 10000, version % 100)
+
+
+def enrich_config_from_running_instance(dsn: str, config: Dict[str, Any], no_value_msg: str) -> None:
+    from patroni.postgresql.config import parse_dsn
+    from patroni.config import _AUTH_ALLOWED_PARAMETERS
+
+    parsed_dsn = parse_dsn(dsn)
+    if not parsed_dsn:
+        raise PatroniCtlException('Failed to parse DSN string')
+    if not parsed_dsn.get('user'):
+        raise PatroniCtlException('No user provided in the DSN')
+
+    from . import psycopg
+    conn = psycopg.connect(dsn=dsn)
+
+    with conn.cursor() as cur:
+        cur.execute("SELECT 1 FROM pg_roles WHERE rolname=%s AND rolsuper='t';", (parsed_dsn['user'],))
+        if cur.rowcount < 1:
+            conn.close()
+            raise PatroniCtlException('The provided user does not have superuser privilege')
+
+        cur.execute("SELECT name, current_setting(name) FROM pg_settings\
+                     WHERE context <> 'internal' \
+                     AND source IN ('configuration file', 'command line', 'environment variable') \
+                     AND category <> 'Write-Ahead Log / Recovery Target' \
+                     AND setting <> '(disabled)' \
+                     OR name IN ('hba_file', 'ident_file', 'config_file', \
+                                 'data_directory', \
+                                 'listen_addresses', 'port', \
+                                 'max_connections', 'max_worker_processes', 'max_wal_senders', \
+                                 'max_replication_slots', 'max_locks_per_transaction', 'max_prepared_transactions', \
+                                 'hot_standby', \
+                                 'wal_level', 'wal_log_hints', \
+                                 'wal_keep_segments', 'wal_keep_size', \
+                                 'track_commit_timestamp');")
+
+        # adjust values
+        for p, v in cur.fetchall():
+            if p == 'data_directory':
+                config['postgresql']['data_dir'] = v
+            elif p in ('archive_command', 'restore_command', 'archive_cleanup_command',
+                       'recovery_end_command', 'ssl_passphrase_command'):
+                # write commands to the local config due to security implications
+                if not config['postgresql']['parameters']:
+                    config['postgresql']['parameters'] = {}
+                config['postgresql']['parameters'][p] = v
+            else:
+                config['bootstrap']['dcs']['postgresql']['parameters'][p] = v
```

Comment:
We don't want to put `'hba_file', 'ident_file', 'config_file'` to the global configuration, as well as `listen_addresses`, `port`, and `cluster_name` to avoid them removing from the structure later.

Reply:
do we want to have `listen_addresses`, `port`, and `cluster_name` in the generated config at all?

Reply:
No, we don't need them in the config.
Although, first two are useful to generate `postgresql.listen`.

Reply:
I guess, this is already done
As well as cluster_name is used to set `scope`